lib/services/mail_fetcher.dart

```diff
@@ -1,73 +1,134 @@
+        if (email.mimeData!.parts!
+            .elementAt(x)
+            .contentType
+            ?.value
+            .toString()
+            .contains("image") ??
+            false) {
+          var attachment = Attachment();
+          attachment.contentID = email.mimeData!.parts!.elementAt(x).headersList!.where((element) => element.name == "Content-ID").first.value.toString().replaceAll('<', '').replaceAll('>', '');
+          attachment.sender = "Test Sender"; //todo: pull from emailBodyHtml by parsing the HTML
+          attachment.attachment = email.mimeData!.parts!
+              .elementAt(x)
+              .decodeMessageData()
+              .toString(); //These are base64 encoded images with formatting
+          attachment.attachmentNoFormatting = attachment.attachment
+              .toString()
+              .replaceAll(
+              "\r\n", ""); //These are base64 encoded images with formatting
+          list.add(attachment);
+        }
+      }
+      return list;
+    } catch (e) {
+      rethrow;
+    }
+  }
+
   /// Retrieve emails based on a start date, sender filter, and subject filter
-  List<String> _getEmails(
-      DateTime startDate, String senderFilter, String subjectFilter) {
-    List<String> emails = <String>[];
-    emails.add(
-        "this should actually be an email object"); //todo: log in and grab actual emails using the provided 3 filters
-    return emails;
+  Future<List<MimeMessage>> _getEmails(DateTime startDate, String username, String password, String senderFilter, String subjectFilter) async {
+    final client = ImapClient(isLogEnabled: true);
+    try {
+      //Retrieve the imap server config
+      var config = await Discover.discover(username, isLogEnabled: false);
+      if (config == null) {
+        return <MimeMessage>[];
+      } else {
+        var imapServerConfig = config.preferredIncomingImapServer;
+        await client.connectToServer(
+            imapServerConfig!.hostname as String, imapServerConfig.port as int,
+            isSecure: imapServerConfig.isSecureSocket);
+        await client.login(username, password);
+        await client.selectInbox();
+
+        String searchCriteria = 'FROM ${senderFilter} SINCE ${_formatTargetDateForSearch(startDate)} SUBJECT "${subjectFilter}"'; //todo: figure out the syntax for "Date After"
+
+        List<ReturnOption> returnOptions = [];
+        ReturnOption option = ReturnOption("all");
+        returnOptions.add(option);
+
+        final searchResult = await client.searchMessages(
+            searchCriteria: searchCriteria);
+        final matchingSequence = searchResult.matchingSequence;
+        if (matchingSequence != null) {
+          return (await client.fetchMessages(matchingSequence, 'BODY.PEEK[]')).messages;
+        }
+        return <MimeMessage>[];
+      }
+    } catch (e) {
+      rethrow;
+    } finally {
+      if (client.isLoggedIn) {
+        await client.logout();
```

Comment:
Couldn't an exception be thrown in `connectToServer`, which would cause a `logout` to get triggered before we logged in? Instead, we could have 3 different functions, `login`, `fetch` and `logout` that get called here. That way we could guarantee that the correct order occurred.